storage/fanout.go

```diff
@@ -282,7 +284,14 @@ func (q *mergeGenericQuerier) Select(sortSeries bool, hints *SelectHints, matche
 		q.setQuerierMap[res.set] = res.qr
 		seriesSets = append(seriesSets, res.set)
 	}
-	return newGenericMergeSeriesSet(seriesSets, q, q.mergeFunc)
+
+	set := &genericMergeSeriesSet{
+		querier:   q,
+		mergeFunc: q.mergeFunc,
+		sets:      seriesSets,
+	}
+	q.mergeSeriesSets = append(q.mergeSeriesSets, set)
```

Comment:
Can you explain here what is happening here? I don't follow this code.. why have to remember the set during `Select` in `mergeSeriesSets`. This sounds wrong as Select can be done multiple times.

Reply:
As @brian-brazil pointed out, we need to make sure all sets from all selects are succeeded https://github.com/prometheus/prometheus/pull/7251#discussion_r426546061 before we merge sets, otherwise we ignore to sets that come from that querier. Thus I added this to remember sets.

Also I've added a comment for this.